### ISSUE TITLE
Proposal: Multi-Device Parallel Optimization for Diffusion Module During Inference

### ISSUE THREAD

**Issue by Amaranterre** [google-deepmind/alphafold3#343]
#### **Overview**

We have identified an opportunity to optimize the **diffusion module** in AlphaFold3 by replacing **`vmap` ([[automatic vectorization](https://docs.jax.dev/en/latest/automatic-vectorization.html)](https://docs.jax.dev/en/latest/automatic-vectorization.html))** with **`shard_map` ( [[manual parallelism](https://docs.jax.dev/en/latest/notebooks/shard_map.html)](https://docs.jax.dev/en/latest/notebooks/shard_map.html) )**. This modification leverages multi-device execution, potentially improving inference speed.

#### **Code Context**

**File:** `alphafold3/src/alphafold3/model/network/diffusion_head.py`
 **Relevant snippet (line 367):**

```
pythonCopyEditinit = (
    jax.random.split(key, num_samples),
    positions,
    jnp.tile(noise_levels[None, 0], (num_samples,)),
)

apply_denoising_step = hk.vmap(
    apply_denoising_step, in_axes=(0, None), split_rng=(not hk.running_init())
)
result, _ = hk.scan(apply_denoising_step, init, noise_levels[1:], unroll=4)
_, positions_out, _ = result
```

Currently, `hk.vmap` is used to apply `apply_denoising_step` across the first axis in a serial manner. However, by replacing `vmap` with `shard_map`, we can enable parallel execution across multiple devices, leading to a significant performance improvement.

------

### **Implementation**

The current approach transforms `apply_denoising_step` via `hk.vmap`, applying it independently along the first axis in a serial manner. By replacing `vmap` with `shard_map`, we achieve a similar functional effect but in a parallelized fashion, utilizing available devices more effectively.

#### **Challenges & Considerations**

1. **Haiku Incompatibility with `shard_map`**
   - Haiku does not natively support `shard_map`, requiring a mix of Haiku and JAX functions.
   - This workaround generated several warnings/errors, which we addressed, but long-term stability remains a concern.
2. **Device Limitations**
   - The first axis length is **only 5**, meaning that at most **5 devices** can be utilized.
   - To address this, we considered **padding** the first axis to enhance scalability.

------

### **Performance Gains**

In our experiments, we tested `shard_map` using **5 NVIDIA RTX 4090 GPUs**, and observed a **50% speed improvement** in the diffusion module, contributing to an overall performance boost.

#### **Potential Issues**

- **Device Limitations:** Since the first axis length is fixed at **5**, the implementation supports a maximum of **5 devices**.
- **Haiku Compatibility:** While we encountered no explicit warnings/errors across **12 test cases**, Haiku's lack of native support for `shard_map` introduces some **uncertainty** regarding long-term stability.

**Comment by Augustin-Zidek** [comment]
However, AlphaFold is optimised for throughput, not latency and as such we recommend people fold multiple inputs (or multiple seeds of the same input) in parallel on multi-GPU setups instead of parallelizing a single prediction across multiple devices.

I will refer people to this issue if they ask similar questions in the future, but I won't change the existing code.

Also, note that only sharding the diffusion head is not optimal, as the pre-diffusion trunk represents the bulk of the compute time for inference.
